### PR TITLE
compute: match environment from line results

### PR DESCRIPTION
**Pull request by rvantonder**
This PR introduces regex submatching and exposes it in match environment. See https://github.com/sourcegraph/sourcegraph/issues/21579 for context.

It is not used anywhere yet. I will hook it up to an experimental GQL or streaming interface in a follow up PR. 

It uses convention of `$1`... names for regexp submatch variables. I will work on named regex capture groups a bit later. This is an early kick off to just get the types right and some basic regex postprocessing.

Larger context: I am minting a new `internal/compute` package for computed values that are typically more expensive to compute than pure text search results. You can generally think of `internal/compute` as "postprocessing logic that manipulates search results". In future I expect that "postprocessing" will include, e.g., hooking into LSIF data _in order to_ still manipulate results. So this package carves out a new place for code that relies on search as a precursor to other result computation, and may make sense to integrate with our worker service and not, e.g., be called from the frontend to do extra result computation.

### CODE REVIEW

File: internal/compute/match.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	"regexp"
+
+	"github.com/sourcegraph/sourcegraph/internal/search/result"
+)
+
+type Location struct {
+	Offset int `json:"offset"`
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
```
Comment (rvantonder) [inline internal/compute/match.go:14]: In my experience it's good to have both line/col and offset data for internally representing location. Interfacing with other tools or APIs means they may natively return (only) offset data, or (only) line/col data. Tools may commit to only one of these representations for performance reasons (e.g., parsing).

It's possible to translate between the two representations--in some contexts, only `offset` or `line/col` is needed. In my experience, when translation is needed, it's more straightforward to unconditionally expose both `line/col` and `offset` and fully hydrate the the location information based on the context in which it's used than requiring the consumer/client to receive one of these and then have to translate it (like `line/col` and then have to call something to get `offset`). Why not always fully hydrate `offset/line/col` and return both? For perf reasons--it's better that we provide the _option_ of hydrating than unconditionally doing work to translate offsets to line/col or  vice versa, if the client only needs one of them.

File: internal/compute/match.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		End:   newLocation(endLine, endColumn, -1),
+	}
+}
+
+func ofRegexpMatches(matches [][]int, lineValue string, lineNumber int) Match {
+	env := make(Environment)
+	var firstValue string
+	var firstRange Range
+	for _, m := range matches {
+		for j := 0; j < len(m); j += 2 {
```
Comment (camdencheek) [inline internal/compute/match.go:59]: A quick note about why we're iterating in sets of 2 might be helpful

File: internal/compute/match.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+			}
+			v := fmt.Sprintf("$%d", j/2)
+			env[v] = Data{Value: value, Range: range_}
+		}
+	}
+	return Match{Value: firstValue, Range: firstRange, Environment: env}
+}
+
+func ofFileMatches(fm *result.FileMatch, r *regexp.Regexp) *Result {
+	var matches []Match
```
Comment (camdencheek) [inline internal/compute/match.go:80]:
```suggestion
	matches := make([]Match, len(fm.LineMatches))
```

**Review by camdencheek — APPROVED** [review]
Code looks good ⚡ 

Out of curiosity, did you look at all into grabbing these subexpression values in the backends rather than re-parsing all the results in a post-processing step? It _probably_ doesn't matter given the relative size of the search corpus and the result sets, and is likely much more difficult, but it feels funny to re-parse all our results.

**Review by coury-clark — APPROVED** [review]

File: internal/compute/match.go (hunk 3 above)
Comment (rvantonder) [reply internal/compute/match.go:80]: Hmmm this adds a `nil` value to matches so I'm inclined to skip preallocating or messing with `len(fm.LineMatches-1)` or some such.

```patch
        --- want                                                                                                                                                                                                                       
        +++ got                                                                                                                                                                                                                        
        @@ -1,6 +1,22 @@                                                                                                                                                                                                               
         `{                                                                                                                                                                                                                            
           "matches": [                                                                                                                                                                                                                
             {                                                                                                                                                                                                                         
        +      "value": "",                                                                                                                                                                                                            
        +      "range": {                                                                                                                                                                                                              
        +        "start": {                                                                                                                                                                                                            
        +          "offset": 0,                                                                                                                                                                                                        
        +          "line": 0,                                                                                                                                                                                                          
        +          "column": 0                                                                                                                                                                                                         
        +        },                                                                                                                                                                                                                    
        +        "end": {                                                                                                                                                                                                              
        +          "offset": 0,                                                                                                                                                                                                        
        +          "line": 0,                                                                                                                                                                                                          
        +          "column": 0                                                                                                                                                                                                         
        +        }                                                                                                                                                                                                                     
        +      },                                                                                                                                                                                                                      
        +      "environment": null                                                                                                                                                                                                     
        +    },                                                                                                                                                                                                                        
        +    {                                                                                                                                                                                                                         
               "value": "abcdefgh",                                                                                                                                                                                                    
               "range": {                                                                                                                                                                                                              
                 "start": {            
```

**Comment by rvantonder** [comment]
> Did you look at all into grabbing these subexpression values in the backends rather than re-parsing all the results in a post-processing step?

I'm pretty sure Zoekt can't supply this which makes it a dealbreaker. Then there's also the comment from the parent issue mentioning [this in slack](https://sourcegraph.slack.com/archives/CHEKCRWKV/p1615571614206400?thread_ts=1615571214.206100&cid=CHEKCRWKV).

I think we could optimize for it with searcher (and comby) but right now this is small potatoes--we just want to get this info out to Code Insights (which is long-running anyway) and for reasons in https://github.com/sourcegraph/sourcegraph/issues/21579 just the fact that we do this in the backend is a win already.

File: internal/compute/match.go (hunk 3 above)
Comment (camdencheek) [reply internal/compute/match.go:80]: do'h. That's what I get for trying to do a suggestion on my phone 🤦 

That should be `matches := make([]Match, 0, len(fm.LineMatches))`

**Comment by rvantonder** [comment]
(FYI this PR @keegancsmith)

**Comment by camdencheek** [comment]
> I'm pretty sure Zoekt can't supply this which makes it a dealbreaker.

Makes sense. Thanks for humoring me 😄

File: internal/compute/match.go (hunk 3 above)
Comment (rvantonder) [reply internal/compute/match.go]: image

File: internal/compute/match.go (hunk 1 above)
Comment (keegancsmith) [reply internal/compute/match.go:23]: you should also make it explicit if it is byte or rune offsets/columns.

**Comment by keegancsmith** [comment]
> I'm pretty sure Zoekt can't supply this which makes it a dealbreaker.

Well we could adjust the API to do this. Relatively straightforward since Zoekt run go's regex engine, so could do it. But the benefit of post-processing is it works across all backends.

**Comment by rvantonder** [comment]
OK. It's not a priority right now, can measure if this is worth doing later.

**Pull request merged**